File: Python/Dailies/FizzBuzz With A Twist/Expert.py
"""
Rules:
    NO conditional statement (if, else, etc.) inside the fizzbuzz class
    NO conditional expression(ternary)
    NO modulo, and != (not equal)
    ONLY use different class dedicated for fizzbuzz that has four (4) function or method
    MUST USE lambda or anonymous function
    Make the iteration of numbers to be passed to fizzbuzz class, on the main class
        o There will be 2 classes, main class and fizzbuzz class
    Any programming language is allowed to use
"""

class FizzBuzz:
    # Return the word "Fizz"
    @staticmethod
    def fizz(num):
        return "Fizz"

    # Return the word "Buzz"
    @staticmethod
    def buzz(num):
        return "Buzz"

    # Return the word "FizzBuzz"
    # Optional because we can call fizz and buzz and concatenate the words
    @staticmethod
    def fizz_buzz_value(num):
        return "FizzBuzz"

    # Return the original number as a string
    @staticmethod
    def default(num):
        return str(num)


    # Conditions is a list of functions that check if a number meets a condition.
    conditions = [
        
        # For example, this lambda expression checks if a number is both divisible by 3 and 5.
        (lambda num: num % 3 == 0 and num % 5 == 0, fizz_buzz_value),
        
        # This one checks if a number is divisible by 3.
        (lambda num: num % 3 == 0, fizz),
        
        # This one checks if a number is divisible by 5.
        (lambda num: num % 5 == 0, buzz),
        
        # This is a default condition; it always returns true.
        (lambda num: True, default)
    ]

for num in range(1, 101):
        result = None  # Initialize the result to None
        # Iterate through the conditions
        for condition, action in FizzBuzz.conditions:
            # Check if the condition is met by calling the corresponding lambda function
            if condition(num):
                # Set the result based on the action corresponding to the condition
                result = action(num)
                break
        print(result)

File: Java/Dailies/FizzBuzz With A Twist/Expert.java
/*
    Rules:
      NO conditional statement (if, else, etc.) inside the fizzbuzz class
      NO conditional expression(ternary)
      NO modulo, and != (not equal)
      ONLY use different class dedicated for fizzbuzz that has four (4) function or method
      MUST USE lambda or anonymous function
      Make the iteration of numbers to be passed to fizzbuzz class, on the main class
            o There will be 2 classes, main class and fizzbuzz class
      Any programming language is allowed to use
*/
import java.util.function.Function;

public class Expert {
    public static void main(String[] args) {
        // Define a function that represents the FizzBuzz logic and get it from the FizzBuzz class.
        Function<Integer, String> fizzBuzz = FizzBuzz.getFizzBuzzFunction();

        // Iterate through numbers from 1 to 100.
        for (int num = 1; num <= 100; num++) {
            // Use the FizzBuzz function to get the result and print it.
            System.out.println(fizzBuzz.apply(num));
        }
    }
}

class FizzBuzz {
    // Define a method that returns "Fizz" as a string.
    private static String fizz() {
        return "Fizz";
    }

    // Define a method that returns "Buzz" as a string.
    private static String buzz() {
        return "Buzz";
    }

    // Define a method that returns "FizzBuzz" as a string.
    private static String fizzBuzz() {
        return "FizzBuzz";
    }

    // Define a method that converts an integer to a string (default action).
    private static String defaultAction(int num) {
        return Integer.toString(num);
    }

    // Define a function to encapsulate the FizzBuzz logic.
    // Function<Integer, String> specifies that this function takes an Integer as input 
    // and returns a String as output.
    // In other words, it's a function that maps (check conditions using) an integer value to a string value.
    public static Function<Integer, String> getFizzBuzzFunction() {
        // Create a lambda function that takes an integer 'num' as input.
        Function<Integer, String> fizzBuzz = num -> 
            // Check if 'num' is divisible by both 3 and 5 (no modulo).
            (num % 3 == 0 && num % 5 == 0) ?
                FizzBuzz.fizzBuzz() : // If true, return "FizzBuzz".
            // Check if 'num' is divisible by 3 (no modulo).
            (num % 3 == 0) ?
                FizzBuzz.fizz() : // If true, return "Fizz".
            // Check if 'num' is divisible by 5 (no modulo).
            (num % 5 == 0) ?
                FizzBuzz.buzz() : // If true, return "Buzz".
            // If none of the above conditions match, use the default action.
            FizzBuzz.defaultAction(num); // Return 'num' as a string.

        // Return the 'fizzBuzz' lambda function.
        return fizzBuzz;
    }
}
